app/controllers/concerns/two_factor_authenticatable.rb

```diff
@@ -33,36 +35,33 @@ def reset_attempt_count_if_user_no_longer_locked_out
     current_user.update(second_factor_attempts_count: 0, second_factor_locked_at: nil)
   end
 
-  def mark_user_session_authenticated
-    user_session[TwoFactorAuthentication::NEED_AUTHENTICATION] = false
-    user_session[:authn_at] = Time.zone.now
-  end
-
   def handle_valid_otp
-    mark_user_session_authenticated
-    bypass_sign_in current_user
-    flash[:notice] = t('devise.two_factor_authentication.success') unless reauthn?
-
-    analytics.track_event('User 2FA successful')
+    if context == 'authentication'
+      handle_valid_otp_for_authentication_context
+    else
```

Comment:
I think that's fine, but I was looking to make sure the code would error if it got a value we didn't specify so it doesn't silently fall through? But I'll defer to you since I don't have much context on the rest of this yet.